Added ability to compile effects at runtime from GLSL

rds1983 opened the pull request:
This PR adds Shader object to the FNA that could be compiled at run-time from GLSL code.
I.e. following is full sample that will draw two textures(second below first), using the new feature:
```c#
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using System.IO;

namespace ImageTest
{
	public class MyGame : Game
	{
		private const string VertexShader = @"
// Attributes
attribute vec4 a_position;
attribute vec4 a_color;
attribute vec2 a_texCoords0;

// Uniforms
uniform mat4 MatrixTransform;

// Varyings
varying vec4 v_color;
varying vec2 v_texCoords;

void main()
{
	v_color = a_color;
	v_texCoords = a_texCoords0;
	gl_Position = MatrixTransform * a_position;
}
";

		private const string FragmentShader = @"
// Uniforms
uniform sampler2D TextureSampler;

// Varyings
varying vec4 v_color;
varying vec2 v_texCoords;

void main()
{
	gl_FragColor = v_color * texture2D(TextureSampler, v_texCoords);
}
";

		private readonly GraphicsDeviceManager graphics;
		private Texture2D _texture1, _texture2;
		private Shader _shader;
		private IndexBuffer _indexBuffer;
		private VertexBuffer _vertexBuffer1, _vertexBuffer2;

		public MyGame()
		{
			graphics = new GraphicsDeviceManager(this);

			IsMouseVisible = true;
			Window.AllowUserResizing = true;
		}

		protected override void LoadContent()
		{
			base.LoadContent();

			// Shader
			_shader = new Shader(GraphicsDevice, VertexShader, FragmentShader,
				new System.Collections.Generic.Dictionary<string, VertexElementUsage>
				{
					["a_position"] = VertexElementUsage.Position,
					["a_color"] = VertexElementUsage.Color,
					["a_texCoords0"] = VertexElementUsage.TextureCoordinate,
				});

			// Textures
			using (var fileStream = File.OpenRead("d:\\temp\\image1.png"))
			{
				_texture1 = Texture2D.FromStream(GraphicsDevice, fileStream);
			}

			using (var fileStream = File.OpenRead("d:\\temp\\image2.png"))
			{
				_texture2 = Texture2D.FromStream(GraphicsDevice, fileStream);
			}

			// Index Buffer
			var indexData = new short[]
			{
				0, 1, 2, 3, 2, 1
			};

			_indexBuffer = new IndexBuffer(GraphicsDevice, IndexElementSize.SixteenBits, indexData.Length, BufferUsage.WriteOnly);
			_indexBuffer.SetData(indexData);

			// Vertex Buffers
			var vertexData1 = new VertexPositionColorTexture[]
			{
				new VertexPositionColorTexture(new Vector3(0, 0, 0), Color.White, new Vector2(0, 0)),
				new VertexPositionColorTexture(new Vector3(_texture1.Width, 0, 0), Color.White, new Vector2(1, 0)),
				new VertexPositionColorTexture(new Vector3(0, _texture1.Height, 0), Color.White, new Vector2(0, 1)),
				new VertexPositionColorTexture(new Vector3(_texture1.Width, _texture1.Height, 0), Color.White, new Vector2(1, 1)),
			};
			_vertexBuffer1 = new VertexBuffer(GraphicsDevice, typeof(VertexPositionColorTexture), vertexData1.Length, BufferUsage.WriteOnly);
			_vertexBuffer1.SetData(vertexData1);

			var vertexData2 = new VertexPositionColorTexture[]
			{
				new VertexPositionColorTexture(new Vector3(0, 0, 0), Color.White, new Vector2(0, 0)),
				new VertexPositionColorTexture(new Vector3(_texture2.Width, 0, 0), Color.White, new Vector2(1, 0)),
				new VertexPositionColorTexture(new Vector3(0, _texture2.Height, 0), Color.White, new Vector2(0, 1)),
				new VertexPositionColorTexture(new Vector3(_texture2.Width, _texture2.Height, 0), Color.White, new Vector2(1, 1)),
			};
			_vertexBuffer2 = new VertexBuffer(GraphicsDevice, typeof(VertexPositionColorTexture), vertexData2.Length, BufferUsage.WriteOnly);
			_vertexBuffer2.SetData(vertexData2);
		}

		protected override void Draw(GameTime gameTime)
		{
			base.Draw(gameTime);

			GraphicsDevice.Clear(Color.Black);

			GraphicsDevice.Indices = _indexBuffer;
			GraphicsDevice.SetVertexBuffer(_vertexBuffer1);

			GraphicsDevice.BlendState = BlendState.NonPremultiplied;

			_shader.Apply();

			Matrix projection;
			Matrix.CreateOrthographicOffCenter(0,
				GraphicsDevice.PresentationParameters.BackBufferWidth,
				GraphicsDevice.PresentationParameters.BackBufferHeight,
				0,
				0,
				-1,
				out projection);

			_shader.SetUniformMatrix("MatrixTransform", ref projection);

			// Draw texture 1
			GraphicsDevice.SetVertexBuffer(_vertexBuffer1);
			GraphicsDevice.Textures[0] = _texture1;
			GraphicsDevice.DrawIndexedPrimitives(PrimitiveType.TriangleList, 0, 0, _vertexBuffer1.VertexCount, 0, _indexBuffer.IndexCount / 3);

			// Draw texture 2 below first
			Matrix m = Matrix.CreateTranslation(0, _texture1.Height + 20, 0) * projection;
			_shader.SetUniformMatrix("MatrixTransform", ref m);
			GraphicsDevice.SetVertexBuffer(_vertexBuffer2);
			GraphicsDevice.Textures[0] = _texture2;
			GraphicsDevice.DrawIndexedPrimitives(PrimitiveType.TriangleList, 0, 0, _vertexBuffer2.VertexCount, 0, _indexBuffer.IndexCount / 3);
		}
	}
}
```

This PR did some refactoring to the OpenGLDevice/ModernGLDevice, described at #210.
Shader.cs code is seriously reworked Shader.java from libgdx(Apache2 license), which is mentioned in the file header.

flibitijibibo commented:
So I'll go ahead and say that I probably won't merge this for two big reasons:

1. Shader.cs is basically someone else's code and has a totally separate license, and that's not something I want to add to FNA's lengthy license already.
2. By default and by design it inserts added work to the core and I don't think there's a way around this.

That said, I'm extremely surprised at how little it actually affects FNA's existing source; at least, a refined version of this would touch very little. The three things this patch screws up:

1. It refactors the devices to depend on one another. I specifically said not to do that and you did it anyway. Undo that.
2. GraphicsDevice and Effect do not need to be changed at all; you could store/check the Shader object in the IGLDevice implementation and do any set/unset actions in, for example, GLDevice.ApplyEffect instead of in Effect.cs.
3. Because this depends on GLSL specifically, you could probably get away with pulling all the new entry points into Shader.cs, and keep all the GLSL work local to that file. More importantly, it reduces the changes to the GLDevices.

i don't think this is a good thing to include in FNA, but I bet this could totally work as an "FNA.GLSL" DLL that injects custom shader management via reflection into GLDevices that use OpenGL. Do whatever you can to minimize the changes to FNA, and see if you can make either a shader override for OpenGLDevice effect functionality at runtime, or introduce a way to register a custom GLDevice that SDL2_FNAPlatform can invoke, and FNA.GLSL can have its own device that's just a ripoff of OpenGLDevice but with GLSL support instead of MojoShader effects.

rds1983 commented:
I created new ShaderGLDevice and moved all logic there. Which allowed to preserve existing code from being changed at all.
Unfortunately, new issue arised.
I've modified above sample to check how well Shader interact with existing Effects framework:
```c#
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using System.IO;

namespace ImageTest
{
	public class MyGame : Game
	{
		private const string VertexShader = @"
// Attributes
attribute vec4 a_position;
attribute vec4 a_color;
attribute vec2 a_texCoords0;

// Uniforms
uniform mat4 MatrixTransform;

// Varyings
varying vec4 v_color;
varying vec2 v_texCoords;

void main()
{
	v_color = a_color;
	v_texCoords = a_texCoords0;
	gl_Position = MatrixTransform * a_position;
}
";

		private const string FragmentShader = @"
// Uniforms
uniform sampler2D TextureSampler;

// Varyings
varying vec4 v_color;
varying vec2 v_texCoords;

void main()
{
	gl_FragColor = v_color * texture2D(TextureSampler, v_texCoords);
}
";

		private readonly GraphicsDeviceManager graphics;
		private Texture2D _texture1, _texture2;
		private Shader _shader;
		private IndexBuffer _indexBuffer;
		private VertexBuffer _vertexBuffer;
		private SpriteBatch _spriteBatch;

		public MyGame()
		{
			graphics = new GraphicsDeviceManager(this);

			IsMouseVisible = true;
			Window.AllowUserResizing = true;
		}

		protected override void LoadContent()
		{
			base.LoadContent();

			_spriteBatch = new SpriteBatch(GraphicsDevice);

			// Shader
			_shader = new Shader(GraphicsDevice, VertexShader, FragmentShader,
				new System.Collections.Generic.Dictionary<string, VertexElementUsage>
				{
					["a_position"] = VertexElementUsage.Position,
					["a_color"] = VertexElementUsage.Color,
					["a_texCoords0"] = VertexElementUsage.TextureCoordinate,
				});

			// Textures
			using (var fileStream = File.OpenRead("d:\\temp\\image1.png"))
			{
				_texture1 = Texture2D.FromStream(GraphicsDevice, fileStream);
			}

			using (var fileStream = File.OpenRead("d:\\temp\\image2.png"))
			{
				_texture2 = Texture2D.FromStream(GraphicsDevice, fileStream);
			}

			// Index Buffer
			var indexData = new short[]
			{
				0, 1, 2, 3, 2, 1
			};

			_indexBuffer = new IndexBuffer(GraphicsDevice, IndexElementSize.SixteenBits, indexData.Length, BufferUsage.WriteOnly);
			_indexBuffer.SetData(indexData);

			// Vertex Buffers
			var vertexData = new VertexPositionColorTexture[]
			{
				new VertexPositionColorTexture(new Vector3(0, 0, 0), Color.White, new Vector2(0, 0)),
				new VertexPositionColorTexture(new Vector3(_texture2.Width, 0, 0), Color.White, new Vector2(1, 0)),
				new VertexPositionColorTexture(new Vector3(0, _texture2.Height, 0), Color.White, new Vector2(0, 1)),
				new VertexPositionColorTexture(new Vector3(_texture2.Width, _texture2.Height, 0), Color.White, new Vector2(1, 1)),
			};
			_vertexBuffer = new VertexBuffer(GraphicsDevice, typeof(VertexPositionColorTexture), vertexData.Length, BufferUsage.WriteOnly);
			_vertexBuffer.SetData(vertexData);
		}

		protected override void Draw(GameTime gameTime)
		{
			base.Draw(gameTime);

//			GraphicsDevice.Clear(Color.Black);

			GraphicsDevice.Indices = _indexBuffer;
			GraphicsDevice.SetVertexBuffer(_vertexBuffer);

			// Draw texture 1
			_spriteBatch.Begin();
			_spriteBatch.Draw(_texture1, Vector2.Zero, Color.White);
			_spriteBatch.End();

			// Draw texture 2 below first
			GraphicsDevice.BlendState = BlendState.NonPremultiplied;
			_shader.Apply();

/*			Matrix projection;
			Matrix.CreateOrthographicOffCenter(0,
				GraphicsDevice.PresentationParameters.BackBufferWidth,
				GraphicsDevice.PresentationParameters.BackBufferHeight,
				0,
				0,
				-1,
				out projection);

			Matrix m = Matrix.CreateTranslation(0, _texture1.Height + 20, 0) * projection;
			_shader.SetUniformMatrix("MatrixTransform", ref m);
			GraphicsDevice.SetVertexBuffer(_vertexBuffer);
			GraphicsDevice.Textures[0] = _texture2;
			GraphicsDevice.DrawIndexedPrimitives(PrimitiveType.TriangleList, 0, 0, _vertexBuffer.VertexCount, 0, _indexBuffer.IndexCount / 3);*/
		}
	}
}
```

Now it draws upper image with SpriteBatch(which means ordinary stock SpriteEffect) and lower image with shader.
Unfortunately, upper image disappears. It happens when device switches to shader in the method ShaderGLDevice.ApplyShader, that has following code:
```c#
		public void ApplyShader(Shader shader)
		{
			if (currentEffect != IntPtr.Zero)
			{
				MojoShader.MOJOSHADER_glEffectEndPass(currentEffect);
				MojoShader.MOJOSHADER_glEffectEnd(currentEffect);
				currentEffect = IntPtr.Zero;
				currentTechnique = IntPtr.Zero;
				currentPass = 0;
			}

			currentShader = shader;
			glUseProgram(shader.ProgramHandle);
			shaderApplied = true;
		}
```

Maybe I do something wrong. Or maybe MojoShader isnt supposed to work well alongside with ordinary shader calls. Anyway, I am closing this PR. It was interesting experiment though.
